File: QAT_KAN/np_forward/Descriptor_builder.py
import numpy as np

# === simple_descriptor_builder ===
def Qmn_descriptor_builder(coords,M_num_bits=7,N_num_bits=20):
    # coords =np.load(coords)
    # 取出各个原子的坐标
    # attention:对应关系取决于output文件的Atomic coordinates (Ang)
    # shape[N,3]
    O  = coords[:,0,:]
    H1 = coords[:,1,:]
    H2 = coords[:,2,:]

    # 利用欧几里得距离公式求出键长--沿着(xyz轴)第一维度求范数，保留样本数N
    # shape(N,)
    r1_sqaure = np.sum(((O - H1) ** 2),axis = 1)
    r2_sqaure = np.sum(((O - H2) ** 2),axis = 1)
    r1 = np.round(np.sqrt(r1_sqaure)).astype(np.int64)
    r2 = np.round(np.sqrt(r2_sqaure)).astype(np.int64)
    # r1 = np.linalg.norm(O - H1, axis=1)  # O-H1 键长
    # r2 = np.linalg.norm(O - H2, axis=1)  # O-H2 键长
    # =========CHECK SHAPE==========
    # print(f"r1:{r1.shape}")
    # print(f"r2:{r2.shape}")
    # 满足对称不变性
    d_mean = (r1 + r2) // 2      # 平均 O–H 键长
    d_diff = np.abs(r1 - r2)      # 键长不对称性（>=0）
    # 键角计算
    v1 = H1 - O
    v2 = H2 - O 
    dot_mul = np.sum(v1 * v2, axis=1)# 点积--对应元素相乘(v1*v2)再求和(sum)
    # np.linalg.norm--计算向量模长
    cos_theta = (dot_mul << N_num_bits) // (r1 * r2 )

    # 将三个描述子按axis=1位置拼接
    descriptor = np.stack([d_mean, d_diff, cos_theta], axis=1) # shape(N)-->shape(N,3)
    return descriptor

File: FPGA/FPGA_Prj/Vitis/ps_pl_axi_test/vitis_workspace/Verlet_NVE_CPU/src/Model/src/Descriptor_Builder.c
#include <math.h>
#include <string.h>

// 三维向量模长计算函数
static inline float vector_norm(const float v[3]) {
    return sqrtf(v[0]*v[0] + v[1]*v[1] + v[2]*v[2]);
}

void simple_descriptor_builder(int N, const float coords[N][3][3], float descriptor[N][3]) {
    //从N个样本中依此提取出O H1 H2原始坐标
    for(int i=0;i<N;i++){
        const float* O  = coords[i][0]; //指向O原子坐标(xyz)
        const float* H1 = coords[i][1]; //指向H1原子坐标(xyz)
        const float* H2 = coords[i][2]; //指向H2原子坐标(xyz)

        //计算OH1和OH2向量模长
        float O_H1[3] = {O[0]-H1[0], O[1]-H1[1], O[2]-H1[2]};
        float O_H2[3] = {O[0]-H2[0], O[1]-H2[1], O[2]-H2[2]};
        float r1 = vector_norm(O_H1);
        float r2 = vector_norm(O_H2); 

        //计算d_mean 和 d_diff
        float d_mean =(r1 + r2) / 2.0f;
        float d_diff = fabsf(r1 - r2);

        //计算cos_theta
        float v1[3] = {H1[0]-O[0], H1[1]-O[1], H1[2]-O[2]};
        float v2[3] = {H2[0]-O[0], H2[1]-O[1], H2[2]-O[2]}; 
        float dot_mul = v1[0]*v2[0] + v1[1]*v2[1] + v1[2]*v2[2];
        float cos_theta = dot_mul / (r1 * r2);

        //output descriptor         shape:(N,3)
        descriptor[i][0] = d_mean;
        descriptor[i][1] = d_diff;
        descriptor[i][2] = cos_theta;
    }
}